pkg/clustermesh/internal/remote_cluster.go

```diff
@@ -0,0 +1,377 @@
+		if err != nil {
+			return nil, fmt.Errorf("failed to detect whether the cluster configuration is required: %w", err)
+		}
+	}
+
+	cfgch := make(chan *types.CiliumClusterConfig)
+	defer close(cfgch)
+
+	// We retry here rather than simply returning an error and relying on the external
+	// controller backoff period to avoid recreating every time a new connection to the remote
+	// kvstore, which would introduce an unnecessary overhead. Still, we do return in case of
+	// consecutive failures, to ensure that we do not retry forever if something strange happened.
+	ctrlname := rc.remoteConnectionControllerName + "-cluster-config"
+	defer rc.controllers.RemoveControllerAndWait(ctrlname)
+	rc.controllers.UpdateController(ctrlname, controller.ControllerParams{
+		DoFunc: func(ctx context.Context) error {
+			rc.getLogger().Debug("Retrieving cluster configuration from remote kvstore")
+			config, err := cmutils.GetClusterConfig(ctx, rc.name, backend)
+			if err != nil {
+				return err
+			}
+
+			if config == nil && requireConfig {
+				return errors.New("cluster configuration expected to be present but not found")
+			}
+
+			// We should stop retrying in case we either successfully retrieved the cluster
+			// configuration, or we are not required to wait for it.
+			cfgch <- config
+			return nil
+		},
+		Context:          ctx,
+		MaxRetryInterval: 30 * time.Second,
+	})
+
+	// Wait until either the configuration is retrieved, or the context expires
+	select {
+	case config := <-cfgch:
+		return config, nil
+	case <-ctx.Done():
+		return nil, fmt.Errorf("failed to retrieve cluster configuration")
+	}
+}
+
+func (rc *remoteCluster) makeEtcdOpts() map[string]string {
+	opts := map[string]string{
+		kvstore.EtcdOptionConfig: rc.configPath,
+	}
+
+	for key, value := range option.Config.KVStoreOpt {
+		switch key {
+		case kvstore.EtcdRateLimitOption, kvstore.EtcdListLimitOption,
+			kvstore.EtcdOptionKeepAliveHeartbeat, kvstore.EtcdOptionKeepAliveTimeout:
+			opts[key] = value
+		}
+	}
+
+	return opts
+}
+
+func (rc *remoteCluster) makeExtraOpts() kvstore.ExtraOptions {
+	return kvstore.ExtraOptions{
+		NoLockQuorumCheck:            true,
+		ClusterName:                  rc.name,
+		ClusterSizeDependantInterval: rc.clusterSizeDependantInterval,
+	}
+}
+
+func (rc *remoteCluster) onInsert() {
+	rc.getLogger().Info("New remote cluster configuration")
+
+	if skipKvstoreConnection {
+		return
+	}
+
+	rc.remoteConnectionControllerName = fmt.Sprintf("remote-etcd-%s", rc.name)
+	rc.restartRemoteConnection()
+
+	go func() {
+		for {
+			val := <-rc.changed
+			if val {
+				rc.getLogger().Info("etcd configuration has changed, re-creating connection")
+				rc.restartRemoteConnection()
+			} else {
+				rc.getLogger().Info("Closing connection to remote etcd")
+				return
+			}
+		}
+	}()
+
+	go func() {
+		for {
+			select {
+			// terminate routine when remote cluster is removed
+			case _, ok := <-rc.changed:
```

Comment:
Please correct me if I am wrong, but I think that this channel receive will race with the receive from line 273. If this goroutine happens to get in `select` while the goroutine above is not executing and there is a message in `rc.changed` channel buffer, it will get the message from channel, check `ok` value to be `true` (since the channel is not closed) and this goroutine will return while `restartRemoteConnection` would not be called above.

To fix that behaviour, I would suggest merging both goroutines into one so that there is only one select.

Reply:
Yes, you are right. It seems to me that this logic is also possibly prone to other race conditions, as I've seen connections being restarted multiple times in a row following a single failure. I'll refactor it make it more robust.

Reply:
9668440e8b98e32cd4a5281e8cd25e689a07bafc addresses this issue refactoring the watchdog logic. I've still maintained the existing `changed` channel logic when an update is detected to avoid introducing too many changes in this PR which is already large due to moving things around.

Reply:
🤔 There's an issue with the new logic. Let me fix it

Reply:
Error cases should be correctly handled now.